lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
-        internal ContextAndServiceConfiguration GetNextConfiguration(string evaluatorId)
+        internal void ReachedActiveContext(string evaluatorId)
+        {
+            lock (_lock)
+            {
+                if (!_submittedEvaluators.Contains(evaluatorId))
+                {
+                    var msg = string.Format("Evaluator:{0} never loaded data but still reached active context stage",
+                        evaluatorId);
+                    Exceptions.Throw(new Exception(msg), Logger);
+                }
+
+                if (_contextLoadedEvaluators.Contains(evaluatorId))
+                {
+                    var msg = string.Format("Evaluator:{0} already reached the active context stage", evaluatorId);
+                    Exceptions.Throw(new Exception(msg), Logger);
+                }
+
+                _contextLoadedEvaluators.Add(evaluatorId);
+                _submittedEvaluators.Remove(evaluatorId);
+            }
+        }
+
+        /// <summary>
+        /// Gets next context configuration. Either master or mapper.
+        /// </summary>
+        /// <param name="evaluatorId">Evaluator Id</param>
+        /// <returns>The context and service configuration</returns>
+        internal ContextAndServiceConfiguration GetNextContextConfiguration(string evaluatorId)
         {
             lock (_lock)
             {
                 if (_submittedEvaluators.Contains(evaluatorId))
                 {
-                    Exceptions.Throw(new Exception("The evaluator is already submitted"), Logger);
+                    string msg = string.Format("The context is already submitted to evaluator:{0}", evaluatorId);
+                    Exceptions.Throw(new Exception(msg), Logger);
                 }
 
-                if (_failedEvaluators.Count == 0 && _assignedPartitionDescriptors >= _numNodes)
+                if (_masterEvaluatorId == null)
+                {
+                    Logger.Log(Level.Info, "Submitting root context and service for master");
+                    _masterEvaluatorId = evaluatorId;
+                    _submittedEvaluators.Add(evaluatorId);
+                    return new ContextAndServiceConfiguration(
+                        ContextConfiguration.ConfigurationModule.Set(ContextConfiguration.Identifier,
+                            IMRUConstants.MasterContextId).Build(),
+                        TangFactory.GetTang().NewConfigurationBuilder().Build());
+                }
+
+                Logger.Log(Level.Info, "Submitting root context and service for a map task");
+                return GetNextDataLoadingConfiguration(evaluatorId);
+            }
+        }
+
+        /// <summary>
+        /// Checks whether the context id belongs to master
+        /// </summary>
+        /// <param name="activeContextId">context id</param>
+        /// <returns>true of it is master context id, false otherwise</returns>
+        internal bool IsMasterContext(string activeContextId)
```

Comment:
In other places, you look these up by the evaluator ID. I am confused as to how we decide which is the master? By Evaluator or Context? Also, shouldn't the group communications topology know this?

If the master is the evaluator / context that runs the update, maybe changing the name to reflect that would help, e.g. `IsIdOfUpdateContext()`?